ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+						// wait for one third of period to try again.
+						thread::sleep(Duration::from_secs(self.period / 3 + 1));
+						return Seal::None;
+					}
+
+					let inturn = state.is_inturn(block.header.number(), &author);
+
+					let now = SystemTime::now();
+
+					let limit = match inturn {
+						true => state.next_timestamp_inturn.unwrap_or(now),
+						false => state.next_timestamp_noturn.unwrap_or(now),
+					};
+
+					// Wait for the right moment.
+					if now < limit {
+						trace!(target: "engine",
+							   "generate_seal: sleeping to sign: inturn: {}, now: {:?}, to: {:?}.",
+							   inturn, now, limit);
+						match limit.duration_since(SystemTime::now()) {
+							Ok(duration) => {
+								thread::sleep(duration);
+							},
+							Err(e) => {
+								warn!(target:"engine", "generate_seal: unable to sleep, err: {}", e);
+								return Seal::None;
+							}
+						}
+					}
+
+					trace!(target: "engine", "generate_seal: seal ready for block {}, txs: {}.",
+						   block.header.number(), block.transactions.len());
+					return Seal::Regular(null_seal);
+				}
+			}
+		}
+		Seal::None
+	}
+
+	fn verify_local_seal(&self, _header: &Header) -> Result<(), Error> { Ok(()) }
+
+	fn verify_block_basic(&self, header: &Header) -> Result<(), Error> {
+		// Largely same as https://github.com/ethereum/go-ethereum/blob/master/consensus/clique/clique.go#L275
+
+		// Ignore genesis block.
+		if header.number() == 0 {
+			return Ok(());
+		}
+
+		// Don't waste time checking blocks from the future
+		{
+			let limit = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs() + self.period;
+			if header.timestamp() > limit {
+				return Err(Box::new(
+					format!("Block is too far in the future, timestamp: {}, limit: {}", header.timestamp(), limit)
+				).into());
+			}
+		}
+
+		let is_checkpoint = header.number() % self.epoch_length == 0;
+
+		if is_checkpoint && *header.author() != NULL_AUTHOR {
+			return Err(Box::new("Checkpoint block must enforce zero beneficiary").into());
+		}
+
+		// Nonce must be 0x00..0 or 0xff..f
+		let nonce = header.decode_seal::<Vec<&[u8]>>()?[1];
+		if *nonce != NONCE_DROP_VOTE && *nonce != NONCE_AUTH_VOTE {
+			return Err(Box::new("nonce must be 0x00..0 or 0xff..f").into());
+		}
+		if is_checkpoint && *nonce != NULL_NONCE[..] {
+			return Err(Box::new("Checkpoint block must have zero nonce").into());
+		}
+
+		// Ensure that the extra-data contains a signer list on checkpoint, but none otherwise.
+		if (!is_checkpoint && header.extra_data().len() != (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH))
+			|| (is_checkpoint && header.extra_data().len() <= (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH))
+			|| (is_checkpoint && (header.extra_data().len() - (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH)) % 20 != 0) {
+			return Err(Box::new(format!("Invalid extra_data length, got {}", header.extra_data().len())).into());
+		}
+
+		// Ensure that the mix digest is zero as we don't have fork protection currently
```

Comment:
This comment makes it seem like Parity doesn't have fork protection, but looking at Geth it seems that Clique in general doesn't have fork protection. Could this be cleared up in the comment?